detectors/gcp/utils_test.go

```diff
@@ -14,62 +14,87 @@
+	"strings"
+
+	"cloud.google.com/go/compute/metadata"
+)
+
+func NewTestDetector(metadataProvider *FakeMetadataProvider, os *FakeOSProvider) *Detector {
+	return &Detector{metadata: metadataProvider, os: os}
 }
 
 type FakeMetadataProvider struct {
-	Err                  error
-	Attributes           map[string]string
-	InstanceAttributes   map[string]string
-	FakeInstanceID       string
-	FakeInstanceName     string
-	FakeInstanceHostname string
-	FakeZone             string
-	Project              string
+	Err      error
+	Metadata map[string]string
 }
 
-func (f *FakeMetadataProvider) ProjectID() (string, error) {
+func (f *FakeMetadataProvider) Get(s string) (string, error) {
 	if f.Err != nil {
 		return "", f.Err
 	}
-	return f.Project, nil
-}
-func (f *FakeMetadataProvider) InstanceID() (string, error) {
-	if f.Err != nil {
-		return "", f.Err
+	value, ok := f.Metadata[s]
+	if ok {
+		return value, nil
 	}
-	return f.FakeInstanceID, nil
+	return "", metadata.NotDefinedError(s)
 }
-func (f *FakeMetadataProvider) Get(s string) (string, error) {
-	if f.Err != nil {
-		return "", f.Err
-	}
-	return f.Attributes[s], nil
+
+// These keys are all documented at https://cloud.google.com/compute/docs/metadata/predefined-metadata-keys
+func (f *FakeMetadataProvider) ProjectID() (string, error) {
+	return f.Get("project/project-id")
+}
+func (f *FakeMetadataProvider) InstanceID() (string, error) {
+	return f.Get("instance/id")
 }
 func (f *FakeMetadataProvider) InstanceName() (string, error) {
-	if f.Err != nil {
-		return "", f.Err
-	}
-	return f.FakeInstanceName, nil
+	return f.Get("instance/name")
 }
 func (f *FakeMetadataProvider) Hostname() (string, error) {
-	if f.Err != nil {
-		return "", f.Err
-	}
-	return f.FakeInstanceHostname, nil
+	return f.Get("instance/hostname")
 }
 func (f *FakeMetadataProvider) Zone() (string, error) {
-	if f.Err != nil {
-		return "", f.Err
+	value, err := f.Get("instance/zone")
+	if err != nil {
+		return value, err
 	}
-	return f.FakeZone, nil
+	// https://github.com/googleapis/google-cloud-go/blob/35d1f813f755c88d6869d993796c5ddf8ac3eea3/compute/metadata/metadata.go#L663
+	return value[strings.LastIndex(value, "/")+1:], nil
 }
 func (f *FakeMetadataProvider) InstanceAttributeValue(s string) (string, error) {
-	if f.Err != nil {
-		return "", f.Err
+	return f.Get(fmt.Sprintf("instance/attributes/%s", s))
+}
+
+const (
+	fakeRegion              = "us-central1"
+	fakeZone                = fakeRegion + "-a"
+	fakeProjectID           = "my-project"
+	fakeProjectNumber       = "1234567890"
+	fakeInstanceID          = "1234567891"
+	fakeInstanceName        = "my-instance"
+	fakeInstanceHostname    = fakeInstanceName + "." + fakeZone + ".c." + fakeProjectID + ".internal"
+	fakeInstanceMachineType = "n1-standard1"
+)
+
+// fmp constructs a FakeMetadatProvider that responds with the basic GCE metadata fields.
+// To add more metadata, pass key-value pairs as arguments.
+func fmp(keyValues ...string) *FakeMetadataProvider {
```

Comment:
nit: i'd prefer newFakeMetadataProvider to the abbreviation for readability.

Reply:
Done